### ISSUE TITLE
Spurious signals at arm_sin_cos_q31 cosine output

### ISSUE THREAD

**Issue by ua1arn** [ARM-software/CMSIS_5#744]
Then I use `arm_sin_cos_q31` in quadrature NCO - spectrum of cosine signal have additional frequencies, Sine output relative clear.
For Illustration two pictures below. Shown NCO outputs at 700 Hz, fs=48 kHz.
![bad](https://user-images.githubusercontent.com/34999007/69295625-7df44700-0c1c-11ea-9821-37e08c0cd98f.png)
![fine](https://user-images.githubusercontent.com/34999007/69295628-7df44700-0c1c-11ea-9459-a787209728db.png)
If I use `arm_sin_q31` and` arm_cos_q31` - no problems at all.

**Comment by ua1arn** [comment]
Code for illustration:
```
typedef uint32_t ncoftw_t;
typedef int32_t ncoftwi_t;
// Convert ncoftw_t to q31 argument for arm_sin_cos_q31
// The Q31 input value is in the range [-1 0.999999] and is mapped to a degree value in the range [-180 179].
#define FTW2_SINCOS_Q31(angle) ((ncoftwi_t) (angle))
// Convert ncoftw_t to q31 argument for arm_sin_q31
// The Q31 input value is in the range [0 +0.9999] and is mapped to a radian value in the range [0 2*PI).
#define FTW2_COS_Q31(angle) ((q31_t) ((((ncoftw_t) angle) + 0x80000000uL) / 2))
#define IV ivqv [0]
#define QV ivqv [1]
typedef struct
{
	FLOAT_t ivqv [2];
} FLOAT32P_t;

static FLOAT32P_t getsincosf(ncoftw_t angle)
{
	FLOAT32P_t v;
	q31_t sincosv [2];
#if 1
	sincosv [0] = arm_sin_q31(FTW2_COS_Q31(angle));
	sincosv [1] = arm_cos_q31(FTW2_COS_Q31(angle));
#else
	arm_sin_cos_q31(FTW2_SINCOS_Q31(angle), & sincosv [0], & sincosv [1]);
	// at index 0 all fine
	// at index 1 with sidetones
#endif
	arm_q31_to_float(sincosv, v.ivqv, 2);
	return v;
}


```

**Comment by christophe0606** [comment]
@ua1arn I was not able to reproduce a similar issue on my side and always get same results from arm_sin_cos_q31 and (arm_sin_q31 , arm_cos_q31).
I am not saying that there are no issues since obviously the implementations of arm_sin_cos_q31 and (arm_sin_q31 , arm_cos_q31) are quite different.

You C code looks right too.

Internally the functions are using the table sinTable_q31.
There is no corruption of this table ? 

Would it be possible for you to share the input pattern ?

**Comment by ua1arn** [comment]
As I known, sinTable_q31 is a same for all functions?
angles generated by sequential calls of this code (not exact, unnesesary parts removed):

```
#define FTWROUND(ftw) ((uint32_t) (ftw))
FLOAT32P_t get_float_aflo(uint_fast8_t pathi)
{
	const ncoftw_t angle = angle_aflo [pathi];
	const FLOAT32P_t v = getsincosf(angle);
	angle_aflo [pathi] = FTWROUND(angle + anglestep_aflo [pathi]);
	return v;
}

```
Initial value of anglestep_aflo claculaed by
`#define FTWAF(freq) (((int_fast64_t) (freq) << NCOFTWBITS) / ARMI2SRATE)`
Where `NCOFTWBITS `= 32
`ARMI2SRATE `= 48000
`freq `may be 700 or (- 700)

**Comment by christophe0606** [comment]
@ua1arn So, here are my latest experiments.
I have generated the angles using your code. The result looks correct (q31 converted to float).

![imga](https://user-images.githubusercontent.com/48906714/69404106-b10e0780-0cfc-11ea-8199-2b85e643e7a9.jpg)

From this, the reference spectrum I am computing is (we may use different normalization for the FFT).

![imgb](https://user-images.githubusercontent.com/48906714/69404107-b1a69e00-0cfc-11ea-9982-f54fbc435de5.jpg)

When I apply your function FTW2_COS_Q31, I am getting the following angle. It looks correct.
![imgc](https://user-images.githubusercontent.com/48906714/69404108-b1a69e00-0cfc-11ea-968f-a523c32bad39.jpg)

For arm_sin_q31, I am getting this spectrum:

![imgd](https://user-images.githubusercontent.com/48906714/69404109-b1a69e00-0cfc-11ea-9af3-9b2171f2c874.jpg)

For arm_sin_cos_q31 (sin part), I am getting:
![imge](https://user-images.githubusercontent.com/48906714/69404112-b23f3480-0cfc-11ea-8267-161cbb5b1d75.jpg)

Both results look the same.

Nevertheless, there is an issue : the values are opposite. So something must be wrong in the angle computations.

![imgf](https://user-images.githubusercontent.com/48906714/69404113-b23f3480-0cfc-11ea-8881-78a6b63d0541.jpg)

I need to understand why the values are opposite as soon as I can find some time.

**Comment by ua1arn** [comment]
> For arm_sin_cos_q31 (sin part), I am getting:
Spectrum problems appear only in **cosine part of result**.
Wrong angle not an issue for demonstration code, But thank you for pointing me to mistype.

**Comment by christophe0606** [comment]
I've just rechecked with cos and I don't see any problem.

**Comment by christophe0606** [comment]
When I compute the error between both signals (arm_cos_q31 and arm_sin_cos_q31 taking the cos), I get a SNR of -6 dB.

I have computed both spectrum with a BlackmannWindows and I indeed see that the spectrum of the arm_cos_q31 is less clean (so opposite of you since you see the problem with arm_sin_cos_q31).

When I switch to an HammingWindow I no more see (visually) any differences.

So I think the problem is not with the arm_sin_cos_q31 but with the window.
